client/src/app/worker/http/http-subscription-polling.ts

```diff
@@ -0,0 +1,130 @@
+        this._active = false;
+        this.lastHash = undefined;
+
+        if (this.abortCtrl !== undefined) {
+            const abortPromise = new Promise(resolver => (this.abortResolver = resolver));
+            setTimeout(this.abortResolver, 5000);
+            this.abortCtrl.abort();
+            await abortPromise;
+            this.abortResolver = undefined;
+        }
+
+        if (this.currentTimeoutResolver !== undefined) {
+            this.currentTimeoutResolver();
+        }
+    }
+
+    private async nextPoll(): Promise<void> {
+        await this.request();
+        if (this.active && this.reopen) {
+            await new Promise<void>(resolve => {
+                const timeout = setTimeout(() => this.currentTimeoutResolver(), POLLING_INTERVAL);
+
+                this.currentTimeoutResolver = () => {
+                    this.currentTimeoutResolver = undefined;
+                    clearTimeout(timeout);
+                    resolve();
+                };
+            });
+        }
+
+        if (this.active && this.reopen) {
+            await this.nextPoll();
+        }
+    }
+
+    private async request() {
+        const headers: any = {
+            'ngsw-bypass': true
+        };
+
+        if (this.endpoint.authToken) {
+            headers.authentication = this.endpoint.authToken;
+        }
+
+        this.abortCtrl = new AbortController();
+
+        const [url, paramString] = this.endpoint.url.split(`?`);
+        const params = new URLSearchParams(paramString);
+        params.set(`longpolling`, `1`);
+
+        const body = new FormData();
+        body.append(`request`, this.endpoint.payload);
+        body.append(`lastpolling`, this.lastHash || null);
+
+        try {
+            const response = await fetch(url + `?` + params.toString(), {
+                signal: this.abortCtrl.signal,
+                method: this.endpoint.method,
+                headers,
+                body
+            });
+
+            try {
+                const formData = await response.clone().formData();
+                this.lastHash = formData.get(`hash`).toString();
+                this.callbacks.onData(formData.get(`data`).toString());
+            } catch (_) {}
+
+            if (!response.ok) {
+                const data = await this.parseNonOkResponse(response);
+                const error = this.parseErrorFromResponse(response, data);
+                if (this.callbacks.onError) {
+                    this.callbacks.onError(error);
+                } else {
+                    this.callbacks.onData(error);
+                }
+                this._active = false;
+            }
+        } catch (e) {
+            if (e.name !== `AbortError`) {
+                throw e;
+            } else {
+                this.reopen = false;
+            }
+        }
+
+        this.abortCtrl = undefined;
+
+        if (this.abortResolver) {
+            this.abortResolver();
+        }
+    }
```

Comment:
This method is too long, it should be possible to split it up to improve readability